### ISSUE TITLE
No refs as opera

### ISSUE THREAD

**Issue by AmitJoki** [vuejs/eslint-plugin-vue#2136]
**What rule do you want to change?**

no-ref-as-operand

**Does this change cause the rule to produce more or fewer warnings?**

More warnings

**How will the change be implemented? (New option, new default behavior, etc.)?**

Pinia is the official store management solution for Vue 3. It provides `storeToRefs` to be able to destructure state values as `Ref<T>` 

From https://github.com/vuejs/eslint-plugin-vue/blob/master/lib/utils/ref-object-references.js#L268 I see that `toRefs()` is considered as a function that returns an object of `Refs`. It should also consider `storeToRefs` function as it does the same thing.

Additionally if at all it is possible, maybe an option could be provided where people could provide the function names that behave like `toRefs()` in their projects and this rule would work on the `Refs` returned from those functions as well.

**Please provide some example code that this change will affect:**

```vue
const { storeToRefs } from 'pinia';
const { loggedIn } = storeToRefs(useAuthStore());
if (loggedIn) { // should produce a warning here
  // ...
}
```

**What does the rule currently do for this code?**

It does not produce a warning.

**What will the rule do after it's changed?**

It should produce a warning when refs returned by `storeToRefs` is being referenced as an operand.

**Comment by anden-akkio** [comment]
I have a similar circumstance where I'm using Vue Query, which exposes things like `const { isLoading } = useQuery(...)` as refs, but ESLint does not report an issue when a `.value` is forgotten. I don't know if TypeScript itself can catch this either; it's a similar situation to promises, where TypeScript doesn't really care if you haven't awaited something before using it in a condition. 

Any clues on how to improve, or to somehow statically catch this? This is such a large potential error vector that we need to somehow check for this.

**Comment by enekesabel** [comment]
I second this, huge error potential. Just ran into a ton of issues as I was extracting some common props into stores.

**Comment by FloEdelmann** [comment]
typescript-eslint's [`no-unncessary-condition` rule](https://typescript-eslint.io/rules/no-unnecessary-condition) will likely catch this case. For eslint-plugin-vue, it's hard to know which variable stores a ref, but we can of course improve the current logic.

PRs welcome that make `no-ref-as-operand`…

1. … handle Pinia's `storeToRefs` function
2. … tap into typescript-eslint's AST to check type information (if available) in conditions, so type `Ref<…>` is recognized

**Comment by enekesabel** [comment]
@FloEdelmann thanks for the quick reaction, it might be a viable solution for a very type-safe project.

I turned on the suggested rule, and it turned out overly sensitive for the project I'm working on. In order to only check for the cases of breaking the no-ref-as-operand rule I did the following workaround:

1. Find and replace all usages of `import { storeToRefs } from 'pinia';` with `import { toRefs } from 'vue'; const storeToRefs = toRefs;`
2. run linter, fix error
3. Find and replace all `import { toRefs } from 'vue'; const storeToRefs = toRefs;` with `import { storeToRefs } from 'pinia';`

Layman's solution, but in my case it did the trick.  Just make sure do this with a clean git history, and check your diffs.
